website/blog/2019-01-23-jest-24-refreshing-polished-typescript-friendly.md

```diff
@@ -0,0 +1,104 @@
+
+Examples:
+
+Failing assertion
+
+![failing assertion](/img/blog/24-assertion-error.png)
+
+Type mismatch
+
+![different types](/img/blog/24-different-types.png)
+
+Mock function not called
+
+![mock functions](/img/blog/24-mock-function.png)
+
+## Fixing old pain
+
+We've fixed a couple of really old issues in this release.
+
+The first one we'd like to highlight is `console.log` statements going missing. Jest intercepts and collects all logs in order to give you a stack trace to them, as well as provide them to reporters so you can handle them however you want. However, this has led to an issue where they have simply been missing in certain edge cases. Luckily for Jest 24, [@spion](https://twitter.com/spion) has [stepped up](https://github.com/facebook/jest/pull/6871) and fixed this issue. Thank you very much!
+
+The second one is an issue where Jest runs out of memory if the difference in serialization of expected and received value has a huge number of insertion changes (either unexpected because of mistake in test or defect in serializer or expected because of temporary failures during test-driven development). [@ittordepam](https://twitter.com/ittordepam) has [replaced](https://github.com/facebook/jest/pull/6961) the previous diffing algorithm with `diff-sequences` package, which should fix this issue because it uses the theoretical minimum amount of memory. It opens up possibility for word-diffs in the future, similar to what [git provides](https://git-scm.com/docs/git-diff#git-diff---word-diffltmodegt). Please see [this PR](https://github.com/facebook/jest/pull/4619) and don't hesitate to reach out if you want to help make that happen!
+
+## Other highlights
+
+- We have some improvements for `globalSetup` and `globalTeardown` as well - code transformation will be applied to them similar to `setupFiles` and they are now supported as part of `projects`.
+- You can [configure](https://github.com/facebook/jest/pull/6143) Jest's snapshot location, this is mainly useful if you are building tools which use Jest in a larger build process.
+- A quirk of Jest's CLI has been that while some flags and options have been camel cased (such as `runInBand`), others have not been (such as `no-cache`). In Jest 24, both are recognized, meaning you can write your CLI arguments however you want.
+- We've renamed `setupTestFrameworkScriptFile` to `setupFilesAfterEnv`, and made it into an array. We hope this will make it more obvious what the options is for. We have plans to further overhaul the configuration in the next major, see the paragraph in the section below.
+- To reduce the amount of magic Jest performs to “just work™”, in this release we decided to drop automatic injection of `regenerator-runtime` for async code to work. It's not always necessary (e.g. in Node 8+) and we believe it's the user's responsibility to handle it, which is already done if you use `@babel/preset-env` and `@babel/runtime`.
```

Comment:
Users would need to use `@babel/preset-env`, `@babel/plugin-transform-runtime`, and `@babel/runtime` for this to work (when using default settings for `@babel/preset-env`).

Related: https://github.com/facebook/jest/pull/7701

So from what I understand users should do one of:

- Use `@babel/polyfill` if they're developing an app.
- Use `@babel/plugin-transform-runtime`/`@babel/runtime` if they're developing a module.
- Use the target option of `@babel/preset-env` if they're developing for a target that supports `async/await`.

Somewhat confusing, and really more of a Babel issue than a Jest issue, although whatever is said here should still be correct even if it doesn't include all of this information.

Reply:
If your app is working with async await, jest will do as well now (checkout the master branch). I don't see a reason to change examples and guides.

Reply:
I'm testing `jest@24.0.0-alpha.15`, if you follow the instructions as is `async/await` will not work. Including `@babel/preset-env` and `@babel/runtime` is not sufficient.

Reply:
I agree with everything you've said. I just think currently the guide/this post are misleading (this post less so than the guide).

I suppose this post doesn't explicitly say that using `@babel/preset-env` will be sufficient without configuration, but it doesn't mention that it needs to be configured.

Assuming someone is setting up a new project, and they need to setup Babel and Jest, if they follow the guide `async/await` will not work, which I don't think is ideal.

Aside from that I think there are a few scenarios where someone might be compiling/running code without an existing babel config, possibly they're using the TypeScript compiler, or they have their Babel config inside of a Webpack config.

Reply:
I can change the PR to use `targets: {node: 'current'}` if you'd prefer. There are pretty much 3 options to make it work, I don't have a strong opinion about which one is best because it depends on the project.

What I have on the PR currently would be good if you're testing a React app, and you want to have your test compilation be close to your production compilation.

> Sure, but that never worked in Jest. We're not removing some automatic transpilation of async-await, we're just removing forcing regenerator-runtime into babel plugin config whether you asked for it or not

Reading this post I might think I can get away with just using `@babel/preset-env` with the defaults, which I can't. There may also be a number of people also upgrading Babel when they update Jest, and this could be a pain point for them.

Reply:
What do you think about this? https://github.com/facebook/jest/pull/7670/commits/cd3addec6d37285b5607bd77dc95ca3b88c86334

Reply:
I like it 👍 . The link to the guide for more info is good.